src/boot/efi/efi-string.c

```diff
@@ -139,6 +139,109 @@ DEFINE_STRCHR(char16_t, strchr16);
+                switch (*p) {
+                case '\0':
+                        /* End of pattern. Check that haystack is now empty. */
+                        return *h == '\0';
+
+                case '\\':
+                        p++;
+                        if (*p == '\0' || *p != *h)
+                                /* Trailing escape or no match. */
+                                return false;
+                        break;
+
+                case '?':
+                        if (*h == '\0')
+                                /* Early end of haystack. */
+                                return false;
+                        break;
+
+                case '*':
+                        /* No need to recurse for consecutive '*'. */
+                        while (*p == '*')
+                                p++;
+
+                        do {
+                                /* Try matching haystack with remaining pattern. */
+                                if (efi_fnmatch_internal(p, h, max_depth - 1))
+                                        return true;
+
+                                /* Otherwise, we match one char here. */
+                                h++;
+                        } while (*h != '\0');
+
+                        /* End of haystack. Pattern needs to be empty too for a match. */
+                        return *p == '\0';
+
+                case '[':
+                        if (*h == '\0')
+                                /* Early end of haystack. */
+                                return false;
+
+                        bool first = true, can_range = true, match = false;
+                        for (;; first = false) {
+                                p++;
+                                if (*p == '\0')
+                                        return false;
+
+                                if (*p == '\\') {
+                                        p++;
+                                        if (*p == '\0')
+                                                return false;
+                                        match |= *p == *h;
+                                        can_range = true;
+                                        continue;
+                                }
+
+                                /* End of set unless it's the first char. */
+                                if (*p == ']' && !first)
+                                        break;
+
+                                /* Range pattern if '-' is not first or last in set. */
+                                if (*p == '-' && can_range && !first && *(p + 1) != ']') {
+                                        p++;
+                                        if (*p == '\\')
+                                                p++;
+                                        if (*p == '\0')
+                                                return false;
+
+                                        match |= (*(p - 2) <= *h && *h <= *p);
```

Comment:
`(p - 2)` seems not correct. How about the case when the pattern include `'\\'`, i.e. `"[a-\\x]"`?